File: src/main/java/org/afi/sample/java/NullCheck.java
package org.afi.sample.java;

import java.util.Objects;
import java.util.Optional;

// Possible, but only at Runtime
public class NullCheck {

	public static void main(String... args) {
		final var fullName = concatName("José", null, "Silva");
		assert fullName.equals("José Silva");

		try {
			concatName("José", null, null);
			assert false;
		} catch (NullPointerException e) {
			assert "lastName is required".equals(e.getMessage());
		}
	}

	public static String concatName(String firstName, String middleName, String lastName) {
		Objects.requireNonNull(firstName, "firstName is required");
		Objects.requireNonNull(lastName, "lastName is required");
		return String.format("%s %s %s", firstName,
				Optional.ofNullable(middleName).orElse(""),
				lastName);
	}
}

File: src/main/java/org/afi/sample/kotlin/NullCheck.kt
package org.afi.sample.kotlin

fun main() {
    val fullName = concatName("José", null, "Silva")
    assert(fullName == "José Silva")

    // concatName("José", null, null) //Don`t compile

    nullAutomaticCasts("foo")
}

fun concatName(firstName: String, middleName: String?, lastName: String) =
    "$firstName ${middleName ?: ""} $lastName"

fun nullAutomaticCasts(inputString: String?) {
    if (inputString != null) {
        val str: String = inputString
        assert(str == inputString)
    } else {
        // val str: String = inputString //Don`t compile
    }
}
